### PR TITLE
G-protein based reshaping

### PR DESCRIPTION
**Pull request by ssepeq**
This PR contains a facility for coupling two proteins, altering the shape of one to conform to the other.

### CODE REVIEW

**Comment by ssepeq** [comment]
> There is a hydrophobic pocket in the cryo-EM model formed by I125, I220, V224, L227 of OR51E2 and L388, L393, L394 of GNAS. As these are not direct contacts between individual aminos but rather a general grouping, there ought to be a way to define such a thing as another form of contact point.

_Originally posted by @ssepeq in https://github.com/primaryodors/primarydock/issues/260#issuecomment-1585667301_

I125 is part of the MAYDRYVAIC motif. Trying using these as ordinary contacts for now.

**Comment by primaryodors** [comment]
So this is working now? Ready to run the Big Three tests and then merge?

Then onward to unit tests?

**Comment by ssepeq** [comment]
I want to at least get this cronned on the local server. Have to test ORs that have an EXR salt bridge, and write a cronnable PHP script.

**Comment by primaryodors** [comment]
There's already a PHP script at `predict/generate_couple.php`, isn't it already set up on the server cron?

**Comment by ssepeq** [comment]
And it's failing on OR1A1 + hGNAL.

```
couple: malloc.c:2379: sysmalloc: Assertion `(old_top == initial_top (av) && old_size == 0) || ((unsigned long) (old_size) >= MINSIZE && prev_inuse (old_top) && ((unsigned long) old_end & (pagesize - 1)) == 0)' failed.
Aborted (core dumped)
```

**Comment by ssepeq** [comment]
This is not working. This is not even close to working. OR1A1 is failing horribly with both GNAS and GNAL.

**Comment by primaryodors** [comment]
Time to put this aside and work on the unit tests.

**Comment by ssepeq** [comment]
😭

**Comment by ssepeq** [comment]
About to start running the cron on the server. OR1A1 is fixed. Switching gears to the unit tests.

**Comment by ssepeq** [comment]
Though the cron has just begun, it has so far identified a handful of OR1A1 agonists that bind better to the G-protein-coupled PDBs than to the inactive PDB. One example:

```
        "geraniol":
        {
            "inactive": 13.386333333333333,
            "version": 1686708796,
            "Actual": "Agonist",
            "hGNAL": -2.7592,
            "hGNAS2": -8.30297
        },
```

Letting it run for several days to collect more data for predictions. As yet, no non-agonists have been processed. Importantly, not all agonists are giving negative energies in the coupled PDBs, though all but one (dihydrojasmone) give better energies for at least one of the coupled PDBs compared to the inactive model. The ability to expand into other G-proteins will be important, and more work is necessary for the docking itself, which is currently using tumble spheres instead of best-binding.

**Comment by primaryodors** [comment]
Remember that the goal is not simply to identify agonists, but to reliably distinguish agonists from non-agonists.

Since the protein coupling was the original point of this PR, and that seems to be working well, let's go ahead and test the big three tests and if that looks good then this can be merged. The precise details of protein coupling can be addressed in future issues and PRs. It's fine to include the new prediction method, since having any up to date method is better than not having one, however further development on predictions should also happen in future PRs.

**Comment by ssepeq** [comment]
The aminos calculate differently on the local machine vs. the server, so it was important to check in the amino acid SDFs and round the similarities in the grid to the nearest 0.02.

🟩...

**Comment by ssepeq** [comment]
🟩🟩🟥

**Comment by ssepeq** [comment]
Molecular "ESP" stopped working with the new conform_molecules() function. Restoring that might fix the OR1A1 test.

**Comment by primaryodors** [comment]
Not thrilled to hear that functionality was lost in the rewrite of the multi-molecule function, but let's put the "ESP" feature on hold for now. After finishing this PR, we can rename that feature to "lookahead" and implement it in a new branch.

The big three tests were working before the code changes in this PR. It would be best to start by going backwards through the git log until the test begins working again, in order to narrow down which code change caused it to fail. Then that change should be manually removed from the latest commit.

**Comment by ssepeq** [comment]
🟩🟩🟩

File: data/gprots_aligned.txt [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+GNA13              MAD---FLPSRSVLSVCFPGCLLTS---------------------GEAEQQRKSKEIDKCLSREKTYVKRLVKILLLGAGESGKSTFLKQMRIIHGQ----D--FDQRAR---------EEFRPTIY--SNVIKGMRVLVDAREKLHIPWGDNSNQQHGDKMMSFDTRAPMAAQGMVETRVFLQYLPAIRALWADSGIQNAYDRRREFQLGESVKYFLDNLDKLGEPDYIPSQQDILLARRPTKGIHEYDFEIKNVPFKMVDVGGQRSERKRWFECFDSVTSILFLVSSSEFDQVLMEDRLTNRLTESLNIFETIVNNRVFSNVSIILFLNKTDLLEEKVQI-VSIKD-YFLEFE---GDPHCL---RDVQKFLVECFRNKRRDQQQKPL-------------YH--HFTTAINTENIRLVFR-DVKDTILHD-NLKQLML-Q-
+GNA14              MAG-----------------CCCL--------------------SAEEKESQRISAEIERQLRRDKKDARRELKLLLLGTGESGKSTFIKQMRIIHGSGYS-D--EDRKGFTKLVYQNIFTAMQAMIRAMDTLRIQYVCEQNKENAQIIREVEVDKVSMLSREQVE----------------------AIKQLWQDPGIQECYDRRREYQLSDSAKYYLTDIDRIATPSFVPTQQDVLRVRVPTTGIIEYPFDLENIIFRMVDVGGQRSERRKWIHCFESVTSIIFLVALSEYDQVLAECDNENRMEESKALFKTIITYPWFLNSSVILFLNKKDLLEEKIMY--SHLISYFPEYT---GPKQDV---RAARDFILKLYQDQNPDKE-KVI-------------YS--HFTCATDTDNIRFVFA-AVKDTILQL-NLREFNL-V-
+                                                                            IEKXLXXEKE
```
Comment (primaryodors) [inline data/gprots_aligned.txt:17]: What is this line for?

File: predict/generate_couple.php [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+$cfgf = "tmp/{$gpcrid}_{$gpid}.cplcfg";
+$fname = "$cpl_dir/$fam/{$gpcrid}_{$gpid}.pdb";
+
+if (!file_exists("$cpl_dir/$fam")) mkdir("$cpl_dir/$fam");
+
+$cfg = <<<heredoc
+
+PROT1 pdbs/$fam/$gpcrid.upright.pdb
+PROT2 pdbs/Gprot/$gpid.pdb
+# TEMPLATE pdbs/OR51/OR51E2.8f76.pdb
```
Comment (primaryodors) [inline predict/generate_couple.php:78]: Let's uncomment this template line.

File: predict/methods_common.php [hunk 3]
```diff
@@ -171,106 +176,138 @@ function process_dock($metrics_prefix = "", $noclobber = false)
...
+                continue;
+            }
+            else if ($pose && $node>=0)
+            {
+                if (preg_match("/[A-Z][a-z]{2}[0-9]{1,4}/", $coldiv[0]))
+                {
+                    $resno = intval(substr($coldiv[0], 3));
+                    $region = intval(bw_from_resno($protid, $resno));
+                    $lmode = "$mode.$region";
+                    if (isset($metrics_to_process["$mode.rgn"]))
```
Comment (primaryodors) [inline predict/methods_common.php:235]: Should document all the possible array keys for `$metrics_to_process`.

File: test/test_51e2.cplcfg
Comment (primaryodors) [inline test/test_51e2.cplcfg:1]: Is this file being used?

**Review by primaryodors — CHANGES_REQUESTED** [review]

**Comment by ssepeq** [comment]
It is giving results no better than chance. I'm turning off the cron.

```
Results for OR1A1:
56 correct predictions (37.33%).
  65 false positives.
  29 false negatives.

Results for OR1A2:
9 correct predictions (47.37%).
  5 false positives.
  5 false negatives.

Results for OR1B1:
2 correct predictions (100%).
  0 false positives.
  0 false negatives.

Results for OR1C1:
3 correct predictions (7.32%).
  24 false positives.
  14 false negatives.

Results for OR1D2:
9 correct predictions (21.95%).
  0 false positives.
  32 false negatives.

Total:
79 correct predictions (31.23%).
  94 false positives.
  80 false negatives.
```

File: data/gprots_aligned.txt (hunk 1 above)
Comment (ssepeq) [reply data/gprots_aligned.txt]: Neglected to remove it.

**Comment by ssepeq** [comment]
Uncommenting the template line is causing the TMR5 and TMR6 regions to pivot too much, breaking contact with the hydrophobic pocket and resulting in clashes with side chain atoms of TMR3.

**Comment by ssepeq** [comment]
Prolines confuse the new connection code. Also, residues are getting strewn somehow.

**Comment by ssepeq** [comment]
🟩🟩🟩

**Comment by ssepeq** [comment]
Getting some really nice docks spot checking receptors with known agonists.

[OR1D2_hGNAL_bourgeonal.dock.txt](https://github.com/primaryodors/primarydock/files/11835220/OR1D2_hGNAL_bourgeonal.dock.txt)
[OR1D2_hGNAL_floralozone.dock.txt](https://github.com/primaryodors/primarydock/files/11835221/OR1D2_hGNAL_floralozone.dock.txt)
[OR2T34_hGNAL_floralozone.dock.txt](https://github.com/primaryodors/primarydock/files/11835223/OR2T34_hGNAL_floralozone.dock.txt)
[OR2AG1_hGNAS2_isoamyl acetate.dock.txt](https://github.com/primaryodors/primarydock/files/11835226/OR2AG1_hGNAS2_isoamyl.acetate.dock.txt)

File: PREDICTION.md [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+# Writing Your Own Prediction Methods
+
+Until an accurate prediction method exists, we encourage you to write your own prediction methods. You can use the
+`method_basic.php` script as a template, and then apply any variations you wish. If you create a method that makes
+predictions with at least 80% accuracy, we'd be delighted to accept your pull request.
+
+Configurable variables include:
+`$dock_metals`          If true, the dock will use PDBs that end in `.metal.pdb` instead of the default `.upright.pdb`.
+`$dock_retries`         The number of times to retry if no poses returned. Each retry increases the energy limit.
+`$bias_by_energy`       TODO: This has not been implemented yet.
```
Comment (primaryodors) [inline PREDICTION.md:48]: This will be important to squeeze in.

File: PREDICTION.md [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+
+Available parameters for `$metrics_to_process` include:
+`BENERG`                Total binding energy between ligand and protein;
+`BENERG.rgn`            Total binding energy between ligand and specific regions of the protein, e.g. TMR6;
+`vdWRPL`                Total van der Waals repulsion (i.e. where atoms are < 4Å apart) ligand to protein;
+`vdWRPL.rgn`            Total vdW repulsion between ligand and regions of protein;
+`POLSAT`                Ligand's sum polar satisfaction, a measure of polar-polar and nonpolar-nonpolar proximity;
+`PCLASH`                Clashes between residues of the protein, minus clashes in the original PDB model;
+
+`ACVTH.TMR1` thru `ACVTH.TMR7`
```
Comment (primaryodors) [inline PREDICTION.md:59]: Can this be made compatible with the `.rgn` syntax?

**Review by primaryodors — APPROVED** [review]

**Pull request merged**